File: src/lib/scala/Slate.Common/src/main/scala/slate/common/query/QueryEncoder.scala
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2015 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */

package slate.common.query

import slate.common.{DateTime, Strings}

object QueryEncoder {


  def convertVal(value:Any): String =
  {
    value match {
      case null               => ""
      case None               => "null"
      case s:Option[Any]      => convertVal(s.orNull)
      case s:String           => toString(s)
      case s:Int              => s.toString
      case s:Long             => s.toString
      case s:Double           => s.toString
      case s:Boolean          => if(s) "1" else "0"
      case s:DateTime         => "'" + s.toStringMySql + "'"
      case _                  => value.toString
    }
  }


  /**
   * ensures the text value supplied be escaping single quotes for sql.
    *
    * @param text
   * @return
   */
  def ensureValue(text:String):String =
  {
    text match {
      case null  => text
      case ""    => text
      case _     => text.replace("'", "''")
    }
  }


  def ensureField(text:String): String =
  {
    text match {
      case null  => ""
      case ""    => ""
      case _     => text.toLowerCase.trim().filter( c => c.isDigit || c.isLetter || c == '_')
                     .foldLeft[String]("")( (a,b) => a.toString + b.toString)
    }
  }


  /**
   * ensures the comparison operator to be any of ( = > >= < <= != is), other wise
   * defaults to "="
    *
    * @param compare
   * @return
   */
  def ensureCompare(compare:String) :String =
  {
    compare match {
      case "="  =>  "="
      case ">"  =>  ">"
      case ">=" => ">="
      case "<"  =>  "<"
      case "<=" => "<="
      case "!=" => "!="
      case "is" => "is"
      case _    => "="
    }
  }


  def toString(value:String): String ={
    val s = QueryEncoder.ensureValue(value.asInstanceOf[String])
    val res = if(Strings.isNullOrEmpty(s)) "''" else "'" + s + "'"
    res
  }
}

File: src/lib/kotlin/slatekit-common/src/main/kotlin/slatekit/common/query/QueryEncoder.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.common.query

import slatekit.common.DateTime


object QueryEncoder {


    fun convertVal(value: Any): String {
        return when (value) {
            Query.Null  -> "null"
            is String   -> toString(value)
            is Int      -> value.toString()
            is Long     -> value.toString()
            is Double   -> value.toString()
            is Boolean  -> if (value) "1" else "0"
            is DateTime -> "'" + value.toStringMySql() + "'"
            is List<*>  -> "(" + value.joinToString(",", transform = Any?::toString) + ")"
            is Array<*> -> "(" + value.joinToString(",", transform = Any?::toString) + ")"
            else        -> value.toString()
        }
    }


    /**
     * ensures the text value supplied be escaping single quotes for sql.
     *
     * @param text
     * @return
     */
    fun ensureValue(text: String): String =
            if (text.isNullOrEmpty()) {
                ""
            }
            else {
                text.replace("'", "''")
            }


    fun ensureField(text: String): String =
            if (text.isNullOrEmpty()) {
                ""
            }
            else {
                text.toLowerCase().trim().filter { c -> c.isDigit() || c.isLetter() || c == '_' || c == '.' }
            }


    /**
     * ensures the comparison operator to be any of ( = > >= < <= != is), other wise
     * defaults to "="
     *
     * @param compare
     * @return
     */
    fun ensureCompare(compare: String): String =
            when (compare) {
                "="      -> "="
                ">"      -> ">"
                ">="     -> ">="
                "<"      -> "<"
                "<="     -> "<="
                "!="     -> "<>"
                "is"     -> "is"
                "is not" -> "is not"
                "in"     -> "in"
                else     -> ensureField(compare)
            }


    fun toString(value: String): String {
        val s = QueryEncoder.ensureValue(value)
        val res = if (s.isNullOrEmpty()) "''" else "'$s'"
        return res
    }
}
